controllers/argocd/statefulset.go

```diff
@@ -1002,22 +1004,52 @@ func updateNodePlacementStateful(existing *appsv1.StatefulSet, ss *appsv1.Statef
 
 // Returns true if a StatefulSet has pods in ErrImagePull or ImagePullBackoff state.
 // These pods cannot be restarted automatially due to known kubernetes issue https://github.com/kubernetes/kubernetes/issues/67250
-func containsInvalidImage(cr *argoproj.ArgoCD, r *ReconcileArgoCD) bool {
-
-	brokenPod := false
+func containsInvalidImage(cr argoproj.ArgoCD, r ReconcileArgoCD) (bool, error) {
 
 	podList := &corev1.PodList{}
-	listOption := client.MatchingLabels{common.ArgoCDKeyName: fmt.Sprintf("%s-%s", cr.Name, "application-controller")}
+	applicationControllerListOption := client.MatchingLabels{common.ArgoCDKeyName: fmt.Sprintf("%s-%s", cr.Name, "application-controller")}
 
-	if err := r.Client.List(context.TODO(), podList, listOption); err != nil {
+	if err := r.Client.List(context.TODO(), podList, applicationControllerListOption, client.InNamespace(cr.Namespace)); err != nil {
 		log.Error(err, "Failed to list Pods")
+		return false, err
+	}
+
+	if len(podList.Items) == 0 {
+		// No pods, no work to do
+		return false, nil
+	}
+
+	if len(podList.Items) != 1 {
+		// There should only be 0 or 1. If this message is printed, it suggests a problem.
+		log.Info("Unexpected number of pods in 'containsInvalidImage' pod list", "podListItems", fmt.Sprintf("%d", len(podList.Items)), "namespace", cr.Namespace)
+		return false, nil
+	}
```

Comment:
Not sure if it holds true that there should be only 0 or 1. In case of redis-ha, we do have three pods. In case of application controller sharding, we can have 1-n pods. Since each pod will run on a different node usually, each pod could have a different image pull states. Then again, the original logic seemed to be broken in that regard already.

Reply:
Yup, as you said, I was just porting over the original logic. My improvement was at least now we log it :smile: